File: 2019/src/main/java/aoc/Day25.java
package aoc;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;

public class Day25 {
    private static final boolean PRINT = false;
    private static final boolean INTERACTIVE = false;

    private static String runCommand(IntcodeComputer computer, String command) {
        if (!command.endsWith("\n")) command += "\n";
        computer.addInput(command);
        computer.run();
        String output = computer.readAsciiOutput();

        if (PRINT) {
            System.out.println("> " + command);
            System.out.println(output);
        }

        return output;
    }

    public long part1(String input) {
        IntcodeComputer computer = IntcodeComputer.fromString(input);
        if (INTERACTIVE) {
            computer.runInteractive();
            return 0;
        }

        computer.setWaitForInput(true);

        String[] commands = {
                "south", "take space law space brochure",
                "south", "take mouse",
                "south", "take astrolabe",
                "south", "take mug",
                "north", "north", "west", "north", "north", "take wreath",
                "south", "south", "east", "north", "west", "take sand",
                "north", "take manifold",
                "south", "west", "take monolith",
                "west",
                "drop space law space brochure",
                "drop monolith",
                "drop manifold",
                "drop mouse",
                "west",
        };

        String response = null;
        for (String command : commands)
            response = runCommand(computer, command);

        String password = response.replaceAll("[^0-9]", "");
        return Long.parseLong(password);
    }

    public static void main(String[] args) throws Exception {
        Day25 day = new Day25();
        String filepath = "src/main/resources/day25.txt";
        String input = Files.readString(Paths.get(filepath), StandardCharsets.UTF_8);

        long startPart1 = System.currentTimeMillis();
        long resultPart1 = day.part1(input);
        long endPart1 = System.currentTimeMillis();
        System.out.println("Part 1: " + resultPart1 + " (" + (endPart1 - startPart1) + " ms)");
    }
}

File: 2020/src/main/kotlin/Day25.kt
class Day25 {

    private val mod = 20201227

    private fun parse(input: String): Pair<Long, Long> = input
        .lines()
        .map { it.toLong() }
        .let { it[0] to it[1] }

    fun part1(input: String): Long = parse(input).let { (card, door) ->
        val subjectNumber = 7L
        var value = 1L
        var loop = 0
        while (value != card) {
            value = (value * subjectNumber) % mod
            loop++
        }
        var encryptionKey = 1L
        repeat(loop) {
            encryptionKey = (encryptionKey * door) % mod
        }
        encryptionKey
    }
}

fun main() {
    val day = Day25()
    val input = readText("day25.txt")

    val startPart1 = System.currentTimeMillis()
    val answerPart1 = day.part1(input)
    val endPart1 = System.currentTimeMillis()
    println("Part 1: $answerPart1 (${endPart1 - startPart1} ms)")
}
